servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapSingle.java

```diff
@@ -222,29 +224,58 @@ private void enqueueAndDrain(Object item) {
             assert s != null;
 
             if (!pending.offer(item)) {
-                QueueFullException exception = new QueueFullException("pending");
-                if (item instanceof TerminalNotification) {
-                    LOGGER.error("Queue should be unbounded, but an offer failed!", exception);
-                    throw exception;
-                } else {
-                    onError0(exception, true, true);
-                }
+                enqueueAndDrainFail(item);
             }
             drainPending(s);
         }
 
         private void drainPending(Subscription subscription) {
-            long drainedCount = drainSingleConsumerQueue(pending, this::sendToTarget, emittingUpdater, this);
-            if (drainedCount != 0) {
+            long drainCount = 0;
+            Throwable delayedCause = null;
+            while (acquirePendingLock(emittingUpdater, this)) {
+                try {
+                    Object t;
+                    while ((t = pending.poll()) != null) {
+                        ++drainCount;
+                        try {
+                            sendToTarget(t);
+                        } catch (Throwable cause) {
+                            if (delayedCause == null) {
+                                delayedCause = cause;
```

Comment:
nit: should we use `addSuppressed(...)` when `delayedCause != null`

Reply:
lets discuss on https://github.com/apple/servicetalk/pull/1011#discussion_r408440046